servers/src/main/java/tachyon/master/block/BlockMaster.java

```
@@ -0,0 +1,571 @@
+   * @param blockId the committing block id
+   * @param length the length of the block
+   */
+  public void commitBlock(long workerId, long usedBytesOnTier, int tierAlias, long blockId,
+      long length) {
+    LOG.debug("Commit block from worker: {}",
+        FormatUtils.parametersToString(workerId, usedBytesOnTier, blockId, length));
+
+    MasterWorkerInfo workerInfo = mWorkers.getFirstByField(mIdIndex, workerId);
+    workerInfo.addBlock(blockId);
+    workerInfo.updateUsedBytes(tierAlias, usedBytesOnTier);
+    workerInfo.updateLastUpdatedTimeMs();
+
+    MasterBlockInfo masterBlockInfo = mBlocks.get(blockId);
+    if (masterBlockInfo == null) {
+      masterBlockInfo = new MasterBlockInfo(blockId, length);
+      mBlocks.put(blockId, masterBlockInfo);
+      // write new block info to journal.
+      writeJournalEntry(
+          new BlockInfoEntry(masterBlockInfo.getBlockId(), masterBlockInfo.getLength()));
+      flushJournal();
+    }
+    masterBlockInfo.addWorker(workerId, tierAlias);
+    mLostBlocks.remove(blockId);
+  }
+
+  /**
+   * Commit a block, but without a worker location. This means the block is only in ufs.
+   *
+   * @param blockId the id of the block to commit
+   * @param length the length of the block
+   */
+  public void commitBlock(long blockId, long length) {
+    LOG.debug("Commit block: {}", FormatUtils.parametersToString(blockId, length));
+
+    MasterBlockInfo masterBlockInfo = mBlocks.get(blockId);
+    if (masterBlockInfo == null) {
+      masterBlockInfo = new MasterBlockInfo(blockId, length);
+      mBlocks.put(blockId, masterBlockInfo);
+      // write new block info to journal.
+      writeJournalEntry(
+          new BlockInfoEntry(masterBlockInfo.getBlockId(), masterBlockInfo.getLength()));
+      flushJournal();
+    }
+  }
+
+  public BlockInfo getBlockInfo(long blockId) throws BlockInfoException {
+    MasterBlockInfo masterBlockInfo = mBlocks.get(blockId);
+    if (masterBlockInfo != null) {
```

It may be cleaner to do an `== null` check and throw exception instead of having a longer `!= null` clause.